### PR TITLE
add script to refine temporally and spatially a kinematic model

### PR DESCRIPTION
**Pull request by Thomas-Ulrich**
This is the script I designed to refine temporally and spatially the kinematic model of Northridge.
Basically create more source points by interpolation, and smooth also the SR.
Needed for the training.

Thomas.

### CODE REVIEW

File: preprocessing/science/refine_srf.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+                if ndt1 == 0:
+                    continue
+                if ndt1 > p1.ndt:
+                    print("this script assumes that ndt is not larger than the first ndt in the file", ndt1, p1.ndt)
+                    raise
+                while True:
+                    line = fid.readline()
+                    lSTF.extend(line.split())
+                    if len(lSTF) == ndt1:
+                        p1.aSR[j, i, 0:ndt1] = np.array([float(v) for v in lSTF])
```
Comment (sebwolf-de) [inline preprocessing/science/refine_srf.py:110]: Here we just read the slip rate in u_1 direction, right? What about the slip rate in u_2 and u_3 direction?

File: preprocessing/science/refine_srf.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        self.strike = np.zeros((ny, nx))
+        self.dip = np.zeros((ny, nx))
+        self.rake = np.zeros((ny, nx))
+        self.aSR = 0
+        self.PSarea = 0
+        self.dt = 0
+        self.myt = 0
+
+    def init_aSR(self):
+        self.aSR = np.zeros((self.ny, self.nx, self.ndt))
```
Comment (sebwolf-de) [inline preprocessing/science/refine_srf.py:31]: Only one sliprate, but according to srf documentation it is possible, that we have also three slip rates per subfault.

File: preprocessing/science/refine_srf.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        aSRa[:, :, k] = scipy.ndimage.zoom(p.aSR[:, :, k], spatial_zoom, order=spatial_order)
+
+    # interpolate temporally the AST
+    for j in range(p2.ny):
+        for i in range(p2.nx):
+            f = interpolate.interp1d(p.myt, aSRa[j, i, :], kind="quadratic")
+            p2.aSR[j, i, :] = f(p2.myt)
+            if p2.slip1[j, i] < 0:
+                p2.aSR[j, i, :] = 0
+                continue
```
Comment (sebwolf-de) [inline preprocessing/science/refine_srf.py:149]: Why do we do this? Is not completly clear to me. Maybe add a comment.

File: preprocessing/science/refine_srf.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                while True:
+                    line = fid.readline()
+                    lSTF.extend(line.split())
+                    if len(lSTF) == ndt1:
+                        p1.aSR[j, i, 0:ndt1] = np.array([float(v) for v in lSTF])
+                        break
+    return p1
+
+
+def upsampleFault(p, spatial_order, spatial_zoom, temporal_zoom):
```
Comment (sebwolf-de) [inline preprocessing/science/refine_srf.py:115]: Could be a member of the FaultPlane.

**Review by sebwolf-de — COMMENTED** [review]
Good morning, looks good overall!
You are not consistent in the use of camelCase and snake_case regarding the naming of methods.

File: preprocessing/science/refine_srf.py (hunk 3 above)
Comment (Thomas-Ulrich) [reply preprocessing/science/refine_srf.py:149]: After the cubic interpolation, slip1 may get negative which does not make sense.

File: preprocessing/science/refine_srf.py (hunk 1 above)
Comment (Thomas-Ulrich) [reply preprocessing/science/refine_srf.py:110]: `Although	the	general	format	given	here	allows	for	slip	in	all	three	directions	(u1, u2, u3),	
most	cases	will	have	only	one	(u1)	or	two	(u1,	u2)	non-zero	slip	components.`
I will add a warning.

**Comment by Thomas-Ulrich** [comment]
Thx Sebastian for the nice review! I think I answered it now.

File: preprocessing/science/refine_srf.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+                        print("this script assumes that dt is the same for all sources", dt, self.dt)
+                        raise
+                    while True:
+                        line = fid.readline()
+                        lSTF.extend(line.split())
+                        if len(lSTF) == ndt1:
+                            self.aSR[j, i, 0:ndt1] = np.array([float(v) for v in lSTF])
+                            break
+
+    def upsample_fault(self, p, spatial_order, spatial_zoom, temporal_zoom):
```
Comment (sebwolf-de) [inline preprocessing/science/refine_srf.py:149]: I'd reverser the order: p2 = p.upsample_fault(...)

**Review by sebwolf-de — COMMENTED** [review]
Good changes: Nice that also v2.0 is now also supported!

File: preprocessing/science/refine_srf.py (hunk 5 above)
Comment (Thomas-Ulrich) [reply preprocessing/science/refine_srf.py]: adressed by last commits?

**Review by sebwolf-de — APPROVED** [review]

**Pull request merged**